File: IdeaProjects/MessageReader/src/main/java/messageReader/NotificationPublisher.java
package messageReader;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.integration.channel.DirectChannel;
import org.springframework.messaging.Message;
import org.springframework.messaging.support.GenericMessage;

import java.util.Objects;

public class NotificationPublisher {


    @Autowired
    @Qualifier("PublisherToOutputChannel")
    public DirectChannel outgoingChannel;

//    public NotificationPublisher(DirectChannel outgoingChannel) {
//        this.outgoingChannel = outgoingChannel;
//    }

    public boolean publish(Notification notification) {
        return outgoingChannel.send(wrapMessage(notification, true));
    }

    public boolean publishResolve(Notification notification) {
        return outgoingChannel.send(wrapMessage(notification, false));
    }

    Message wrapMessage(Notification notification, boolean active) {
        Message<Notification> outgoingMessage;
        notification.setActive(active);
        notification.setEndProcessingTime(Objects.toString(System.nanoTime()));
        outgoingMessage = new GenericMessage<>(notification);
        return outgoingMessage;
    }
}

File: IdeaProjects/ScalaTutorial/ScalaMessageReader/src/main/scala/scalaMessageReader/NotificationPublisher.scala
package scalaMessageReader

import org.springframework.beans.factory.annotation.{Autowired, Qualifier}
import org.springframework.integration.channel.DirectChannel
import org.springframework.messaging.support.GenericMessage


class NotificationPublisher (){

  @Autowired @Qualifier("PublisherToOutputChannel") var outgoingChannel : DirectChannel = _

  var genericMessage : GenericMessage[Notification] = _

  def publish (notification : Notification) : Boolean = {
    sendMessage(notification, active = true)
  }

  def publishResolve (notification: Notification) : Boolean = {
    sendMessage(notification, active = false)
  }

  def sendMessage (notification: Notification, active : Boolean) : Boolean = {
    notification.isActive = active
    notification.endProcessingTime = System.nanoTime().toString
    genericMessage = new GenericMessage[Notification](notification)
    outgoingChannel.send(genericMessage)
  }
}
